Use a NLP api to rule out some suggestions

yoavf opened the pull request:
Try the [Rosette API](https://developer.rosette.com/features-and-functions) to rule out some suggestions.

Example case: prevent `Update Plugin` from being suggested for `Plugin Updated`

yoavf committed:
use a morphology api to rule out some suggestions

yoavf committed:
add caching and logging

akirk commented on includes/db-adapter.php:
```diff
@@ -93,6 +105,69 @@ private function get_suggested_replacements( $entry ) {
 	}
 
+	private function morphology_matches( $entry, $suggestion ) {
+		$entry_morphology = $this->rosette_api_morphology( $entry->singular );
```
This could be cached / taken out of the function and not regenerated in each loop.

akirk replied:
Oh, I see that the underlying function does the caching already. Disregard.

yoavf replied:
Yeah, since we compare the same original to multiple suggestions, it's better to cache each one separately

akirk commented on includes/db-adapter.php:
```diff
@@ -93,6 +105,69 @@ private function get_suggested_replacements( $entry ) {
+			// Does that lemma also exist in our suggestion?
+			if ( false !== $suggestion_key = array_search( $lemma, $suggestion_morphology->lemmas, true ) ) {
+				// Is it the same part of speech?
+				if ( $suggestion_morphology->posTags[ $suggestion_key ] !== $entry_morphology->posTags[ $entry_key ] ) {
```
Probably we should ignore things like `PUNCT`.

yoavf replied:
Actually not necessary, since we start with the lemma - so if we have a punctuation (ex: `.` ) in both strings, the POS will always be the same.

akirk commented:
Nice and quite straightforward. Calls for unit tests :)

yoavf commented:
I'm probably not going to drop the suggestions completely on mismatch (like it is right now), just lower the score to some extent. Building a test suite of some sorts.

yoavf committed:
morphology: ignore for common words

yoavf committed:
don't exclude morphology matches for now, only log them

yoavf committed:
don't run morphology match on suggestions with more than 3 significant words

yoavf committed:
handle possible empty/incomplete responses

yoavf commented:
(rebased on master)

yoavf committed:
refactor to pass mismatch to log action

yoavf committed:
replace morphology test witn syntax dependencies test

Feels much more effective so far

yoavf commented:
Updated to use the syntax dependencies test.

A few tests surfaced that if the root or objects words (as extracted from the original string) are missing from the suggestion, it is safe to drop the suggestion.

I'll try and add some unit tests - but it's promising :)

yoavf committed:
combine syntax and modified morphology tests

Instead of checking all POS, morphology now only looks for nouns and verbs.
Combine that with syntax deps check for "root" and objects.
Use scoring

yoavf committed:
remove debug

yoavf commented:
The last changes aren't live yet. Actually seems a bit redundant now to have both. Maybe morphology will be enough after all.

yoavf committed:
whitespace

yoavf committed:
reduce code duplication

akirk commented on includes/gh-adapter.php:
```diff
@@ -259,7 +259,7 @@ private function pr_suggestion_comment( $diff, $string, $file, $status_page_url
 			$search_in = preg_replace( '/([\'|"]\s?\+_N_\+?\s+[\'|"])/', '', wp_unslash( $search_in ) );
 
-			if ( ! gp_in( $string['singular'], $search_in )  ) {
+			if ( ! preg_match( '/[\'"]' . preg_quote( $string['singular'], '/' ) . '[\'"]/', $search_in ) ) {
```
You could also add `\b` for word boundaries.

yoavf committed:
Another pass at NLP tests
To improve scoring, we now reduce the impact of a word that was flagged missing both by morphology and by syntax
We also do a reverse match, to find words in the suggestion that aren't in the original string

yoavf committed:
replace placeholders to improve NLP results. Strings are replaced with 'Placeholder', numbers with '42' and tags are removed

yoavf committed:
When creating the array of words that will be used to check if a word is missing from the suggestion/original
make sure contractions are seperate words (that's how they are treated by the api).

yoavf committed:
remove end of sentence punctuation

yoavf committed:
fix double negative score, change penalties

yoavf committed:
also trim ellipsis

yoavf committed:
don't suggest string with no translations

yoavf committed:
use the first pr_po as the po to begin with, otherwise we'll miss the headers
